tiledb/sm/query/sparse_global_order_reader.cc

```diff
@@ -0,0 +1,918 @@
+
+  result_tiles_[f].emplace_back(f, t, domain);
+  memory_used_result_tiles_ += sizeof(ResultTile);
+
+  if (memory_used_result_tiles_ > per_fragment_memory) {
+    *budget_exceeded = true;
+    return Status::Ok();
+  }
+  return Status::Ok();
+}
+
+Status SparseGlobalOrderReader::populate_result_cell_slab() {
+  // For easy reference.
+  auto fragment_num = fragment_metadata_.size();
+  auto domain = array_schema_->domain();
+  auto dim_num = array_schema_->dim_num();
+
+  // Get the number of fragments to process.
+  unsigned num_fragments_to_process = 0;
+  for (auto all_loaded : all_tiles_loaded_)
+    num_fragments_to_process += !all_loaded;
+
+  // Add 2 "fragments" to account for result cell slabs and result tiles.
+  num_fragments_to_process += 2;
+
+  auto per_fragment_memory =
+      memory_budget_ * memory_budget_ratio_coords_ / num_fragments_to_process;
+
+  // TODO Spew warning if memory budget is too low for coordinate tiles.
+
+  // Create result tiles.
+  bool tiles_found = false;
+  bool done_loading_coord_tiles = true;
+  if (subarray_.is_set()) {
+    // Load as many tiles as the memory budget allows.
+    for (unsigned int f = 0; f < fragment_num; f++) {
+      uint64_t t = 0;
+      auto tile_num = fragment_metadata_[f]->tile_num();
+      for (t = frag_tile_idx_[f].first; t < tile_num; t++) {
+        // Use the bit vector to only load tiles in the subarray.
+        auto idx = t / 8;
+        auto bit = 1 << (t % 8);
+        if (result_tile_set_[f][idx] & bit) {
+          bool budget_exceeded = false;
+          RETURN_NOT_OK(load_coord_tile(
+              dim_num, per_fragment_memory, f, t, domain, &budget_exceeded));
+          tiles_found = true;
+
+          if (budget_exceeded)
+            break;
+        }
+      }
+
+      all_tiles_loaded_[f] = t == tile_num;
+      done_loading_coord_tiles &= all_tiles_loaded_[f];
+    }
+  } else {
+    // Load as many tiles as the memory budget allows.
+    for (unsigned int f = 0; f < fragment_num; f++) {
+      uint64_t t = 0;
+      auto tile_num = fragment_metadata_[f]->tile_num();
+      for (t = frag_tile_idx_[f].first; t < tile_num; t++) {
+        bool budget_exceeded = false;
+        RETURN_NOT_OK(load_coord_tile(
+            dim_num, per_fragment_memory, f, t, domain, &budget_exceeded));
+        tiles_found = true;
+
+        if (budget_exceeded)
+          break;
+      }
+
+      all_tiles_loaded_[f] = t == tile_num;
+      done_loading_coord_tiles &= all_tiles_loaded_[f];
+    }
+  }
+
+  done_loading_coord_tiles_ = done_loading_coord_tiles;
+
+  // No tiles found, return.
+  if (!tiles_found) {
+    return Status::Ok();
+  }
+
+  // Populate the temp vector for the common functions to work without change.
+  tmp_result_tiles_.clear();
+  for (auto& frag_result_tile : result_tiles_)
+    for (auto& result_tile : frag_result_tile)
+      tmp_result_tiles_.push_back(&result_tile);
+
+  // Read and unfilter zipped coordinate tiles. Note that
```

Comment:
I think we should separate everything above into its own function, otherwise the stats will be all over the place. Let's split the result tile preparation from the actual reading/loading.